test/fgpu/test_send.py

```diff
@@ -0,0 +1,285 @@
+        return make_streams(
+            output_name=NAME,
+            thread_pool=spead2.ThreadPool(1),
+            endpoints=expected_endpoints,
+            interfaces=list(queues.keys()),
+            ttl=4,
+            ibv=False,
+            packet_payload=8192,
+            comp_vector=0,
+            buffer=65536,
+            bandwidth=1e9,
+            send_rate_factor=1.05,
+            feng_id=FENG_ID,
+            num_ants=64,
+            n_data_heaps=N_CHUNKS * N_FRAMES * N_SUBSTREAMS,
+            chunks=chunks,
+        )
+
+
+@pytest.fixture
+def recv_streams(queues: dict[str, list[spead2.InprocQueue]]) -> dict[str, list[spead2.recv.asyncio.Stream]]:
+    """Create streams for receiving the transmitted data."""
+    config = spead2.recv.StreamConfig()
+    streams: dict[str, list[spead2.recv.asyncio.Stream]] = {}
+    for iface in queues.keys():
+        streams[iface] = []
+        for queue in queues[iface]:
+            tp = spead2.ThreadPool(1)
+            stream = spead2.recv.asyncio.Stream(tp, config)
+            stream.add_inproc_reader(queue)
+            streams[iface].append(stream)
+    return streams
+
+
+@pytest.fixture
+def sensors() -> aiokatcp.SensorSet:
+    """Create sensors that the send code updates."""
+    sensors = aiokatcp.SensorSet()
+    for pol in range(N_POLS):
+        sensors.add(
+            aiokatcp.Sensor(
+                int,
+                f"{NAME}.input{pol}.feng-clip-cnt",
+                "Number of output samples that are saturated",
+            )
+        )
+    return sensors
+
+
+def _fill_random(data: np.ndarray, rng: np.random.Generator) -> None:
+    """Fill an array with random bytes.
+
+    The underlying bytes are filled, rather than assuming any particular dtype.
+    """
+    view = data.view(np.uint8)
+    view[:] = rng.integers(0, 256, view.shape, dtype=np.uint8)
+
+
+def test_bad_substreams():
+    """Test that :class:`Chunk` raises an exception if n_substreams doesn't divide n_channels.
+
+    This is really just for code coverage; nothing really depends on it.
+    """
+    dtype = gaussian_dtype(8)
+    with pytest.raises(ValueError):
+        Chunk(
+            np.zeros((N_FRAMES, N_CHANNELS, N_SPECTRA_PER_HEAP, N_POLS), dtype),
+            np.zeros((N_FRAMES, N_POLS), np.uint32),
+            n_substreams=5,
+            feng_id=FENG_ID,
+            spectra_samples=SPECTRA_SAMPLES,
+        )
+
+
+async def test_send(
+    sample_bits: int,
+    queues: dict[str, list[spead2.InprocQueue]],
+    send_streams: list["spead2.send.asyncio.AsyncStream"],
+    recv_streams: dict[str, list[spead2.recv.asyncio.Stream]],
+    chunks: list[Chunk],
+    sensors: aiokatcp.SensorSet,
+    time_converter: TimeConverter,
+) -> None:
+    """Test sending data via the :class:`katgpucbf.fgpu.send.Chunk` interface.
+
+    .. todo::
+
+       This is a very limited test that does not cover sensors, metrics or
+       missing data.
```

Comment:
Either
* this `todo` needs to be updated (Re: `# Check the sensors and Prometheus metrics` at the end here), or
* I don't actually know what it means to test those things.

Both can be true, I guess.

Reply:
It looks like you're right, Amish, I do see checks for sensors and metrics down at the bottom, and missing frames in the next commit (I'm looking at this commit-by-commit).
Probably the TODO needs to be updated.

Reply:
Indeed, it is just a stale TODO.